src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,286 @@
+
+    WebSearchHandler *webSearchHandler = m_searches[id];
+    if (!m_searches.contains(id)) {
+        throw APIError(APIErrorType::BadParams);
+    }
+
+    webSearchHandler->downloadTorrent(torrentUrl, siteUrl);
+}
+
+void SearchController::categoriesAction()
+{
+    QStringList categories;
+    const QString name = params()["name"].trimmed();
+
+    categories << SearchPluginManager::categoryFullName("all");
+    foreach (const QString &category, SearchPluginManager::instance()->getPluginCategories(name))
+        categories << SearchPluginManager::categoryFullName(category);
+
+    setResult(categories.join(","));
+}
+
+void SearchController::pluginsAction()
+{
+    const QStringList allPlugins = SearchPluginManager::instance()->allPlugins();
+    setResult(QJsonArray::fromVariantList(getPluginsInfo(allPlugins)));
+}
+
+void SearchController::installPluginAction()
+{
+    checkParams({"sources"});
+
+    const QStringList sources = params()["sources"].split('|');
+    for (const QString &source : sources)
+        SearchPluginManager::instance()->installPlugin(source);
+}
+
+void SearchController::uninstallPluginAction()
+{
+    checkParams({"names"});
+
+    const QStringList names = params()["names"].split('|');
+    for (const QString &name : names)
+        SearchPluginManager::instance()->uninstallPlugin(name.trimmed());
+}
+
+void SearchController::enablePluginAction()
+{
+    checkParams({"names", "enable"});
+
+    const QStringList names = params()["names"].split('|');
+    const bool enable = parseBool(params()["enable"].trimmed(), false);
+
+    for (const QString &name : names)
+        SearchPluginManager::instance()->enablePlugin(name.trimmed(), enable);
+}
+
+void SearchController::updatePluginsAction()
+{
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+
+    connect(pluginManager, &SearchPluginManager::checkForUpdatesFinished, this, &SearchController::checkForUpdatesFinished);
+    // connect(pluginManager, &SearchPluginManager::checkForUpdatesFailed, this, &SearchController::checkForUpdatesFailed);
+    pluginManager->checkForUpdates();
+}
+
+void SearchController::checkForUpdatesFinished(const QHash<QString, PluginVersion> &updateInfo)
+{
+    if (updateInfo.isEmpty()) {
+        qDebug() << "All your plugins are already up to date.";
+        return;
+    }
+
+    SearchPluginManager *pluginManager = SearchPluginManager::instance();
+    foreach (const QString &pluginName, updateInfo.keys()) {
+        qDebug() << "Updating plugin" << pluginName;
+        pluginManager->updatePlugin(pluginName);
+    }
+}
+
+QString SearchController::generateSearchId() const
+{
+    QString id;
+
+    do {
+        const size_t size = 2;
+        quint32 tmp[size];
+
+        for (size_t i = 0; i < size; ++i)
+            tmp[i] = Utils::Random::rand();
+
+        id = QByteArray::fromRawData(reinterpret_cast<const char *>(tmp), sizeof(quint32) * size).toBase64();
+    }
+    while (m_searches.contains(id));
```

Comment:
I think this is wrong (in webapplication.cpp too), the `do while()` should be replaced by `if()`